src/gtk/toplevel.cpp

```diff
@@ -339,6 +343,21 @@ gtk_frame_configure_callback( GtkWidget*,
 
 void wxTopLevelWindowGTK::GTKConfigureEvent(int x, int y)
 {
+#ifdef __WXGTK3__
+    // First of all check if our DPI has changed.
+    const auto newScaleFactor = GetContentScaleFactor();
+    if ( newScaleFactor != m_scaleFactor )
+    {
+        const auto oldScaleFactor = m_scaleFactor;
+
+        // It seems safer to change it before generating the events to avoid
+        // any chance of reentrancy.
+        m_scaleFactor = newScaleFactor;
+
+        WXNotifyDPIChange(oldScaleFactor, newScaleFactor);
```

Comment:
This is going to fire when a TLW is created on a scaled display, even though it's not really a "change". Not sure if that is your intention.

Reply:
Yes, this is definitely my intention and, in fact, the main motivation for this change: without it, the (generic) `wxTreeCtrl` in the sample is broken on high DPI display because it uses scale factor of 1 which is in effect when it's being initialized and its images are set. Getting this event makes everything work correctly, even if not optimally -- but I have no idea how could we get the correct scale factor before showing the window and I'm not sure if it's even possible.

Reply:
GTK will make a guess for the scale of a window that is not realized yet, which presumably would always be correct for a single monitor or multiple monitors with the same scale, and if wrong would be corrected here. So I think adding `m_scaleFactor = GetContentScaleFactor();` to the end of `Create()` would avoid the extra event most of the time.

Reply:
Thanks, you're absolutely right and I should have realized that the fact that it doesn't work in my case (I have no idea why does GTK consider my 100% DPI monitor primary even if I have the main Gnome stuff on my 200% monitor) doesn't mean that it never works -- in fact it probably does in the vast majority of cases. I've implemented your suggestion now, thanks again.